cpp/include/cuml/neighbors/knn.hpp

```diff
@@ -210,6 +231,27 @@ void knn_regress(raft::handle_t& handle,
                  size_t n_query_rows,
                  int k);
 
+/**
+ * @brief Weighted KNN regression using pre-computed weights
+ *
+ * @param[in] handle RAFT handle
+ * @param[out] out output array on device (size n_samples)
+ * @param[in] knn_indices array on device of knn indices (size n_samples * k)
+ * @param[in] weights pre-computed weight array on device (size n_samples * k)
+ * @param[in] y array of labels on device (size n_samples)
+ * @param[in] n_index_rows number of vertices in index (eg. size of each y array)
+ * @param[in] n_query_rows number of samples in knn_indices and out
+ * @param[in] k number of nearest neighbors in knn_indices
+ */
+void knn_regress_weighted(raft::handle_t& handle,
```

Comment:
Instead of adding additional signatures (which require duplicate docstrings and additional hoops to wrap in cython), increasing the total LOC needed, could we instead add `sample_weight` as an additional argument to the existing APIs (where a value of `nullptr` signifies unweighted)? This is how other APIs handle this (svm, linear svm, ...) and would be my preference.

We currently don't view our C++ APIs as public, they're just an implementation detail, so there's no breaking changes to worry about here. If needed you could add them as a trailing arg with a default value of `nullptr` to minimize churn in other consumers (gtests, bench, :shrug:).

Same comment everywhere new public duplicate weighted APIs are added.